Add signal guard for schedule_iso minion tasks

b10n1k opened the pull request:
Similar to cd99074 to prevent `Job terminated unexpectedly` on results. This ensures that the minion job will retry in case of interruptions, like network glitch in the system, without kill the process.

https://progress.opensuse.org/issues/108983

Martchus reviewed (changes requested):
This change is actually trivial enough that you don't have to test it locally. (When I introduced `OpenQA::Task::SignalGuard` I tested its behavior locally and added unit tests.)

The only problem that comes to mind is receiving `SIGTERM` right after the database transaction creating all the jobs has been concluded. Then we would retry the Minion job and schedule all those jobs again.

It would make sense to call `$ensure_task_retry_on_termination_signal_guard->retry(0)` after the transaction so the Minion job can just finish in this case (which is hopefully not too time consuming). Unfortunately, "after the transaction" means within `_create_jobs_in_database` so one needed to pass down the guard object there.

If everything that comes after the transaction seems to be too time consuming for going on with it we could also call `$ensure_task_retry_on_termination_signal_guard->abort(1)` after the transaction instead. This would leave the remaining work undone (e.g. emitting events) but would guarantee a quick exit without scheduling jobs twice.

b10n1k commented:
> This change is actually trivial enough that you don't have to test it locally. (When I introduced `OpenQA::Task::SignalGuard` I tested its behavior locally and added unit tests.)
> 
> The only problem that comes to mind is receiving `SIGTERM` right after the database transaction creating all the jobs has been concluded. Then we would retry the Minion job and schedule all those jobs again.
> 
> It would make sense to call `$ensure_task_retry_on_termination_signal_guard->retry(0)` after the transaction so the Minion job can just finish in this case (which is hopefully not too time consuming). Unfortunately, "after the transaction" means within `_create_jobs_in_database` so one needed to pass down the guard object there.
> 
> If everything that comes after the transaction seems to be too time consuming for going on with it we could also call `$ensure_task_retry_on_termination_signal_guard->abort(1)` after the transaction instead. This would leave the remaining work undone (e.g. emitting events) but would guarantee a quick exit without scheduling jobs twice.

check https://github.com/os-autoinst/openQA/commit/cbe26b94889c19ac0edf44d53819e7c7e5a00b3c
I will expand some thoughts here for reviewer:

💡likely the change in t/lib/OpenQA/Test/Utils.pm is not mandatory. I thought it will just not make it confusing when you compare with the original function. But other than that it is not really useful

💡I dont know if I should go the next step to add the guard also in API:V1:Iso. it could be another commit but maybe the smaller this one is the better

Martchus commented on t/39-scheduled_products-table.t:
```diff
@@ -9,8 +9,10 @@ use lib "$FindBin::Bin/lib", "$FindBin::Bin/../external/os-autoinst-common/lib";
+use Test::MockObject;
 use OpenQA::Jobs::Constants;
 use OpenQA::Schema::Result::ScheduledProducts qw(ADDED SCHEDULED SCHEDULING CANCELLING CANCELLED);
+use OpenQA::Task::SignalGuard;
```
Please use some kind of fake object in *all* these tests. We probably don't want to influence the signal handling of unit tests.

Martchus commented on lib/OpenQA/Schema/Result/ScheduledProducts.pm:
```diff
@@ -319,7 +319,7 @@ Internal function to actually schedule the ISO, see schedule_iso().
 
 sub _schedule_iso {
-    my ($self, $args) = @_;
+    my ($self, $args, $guard) = @_;
```
I recommend using signatures in new code.

Wabri replied:
```suggestion
```

Wabri commented on lib/OpenQA/Schema/Result/ScheduledProducts.pm:
```diff
@@ -319,7 +319,7 @@ Internal function to actually schedule the ISO, see schedule_iso().
 =cut
 
 sub _schedule_iso {
```
```suggestion
sub _schedule_iso ($self, $args, $guard) {
```

Wabri commented on lib/OpenQA/Task/Iso/Schedule.pm:
```diff
@@ -37,12 +38,14 @@ sub _download_scenario_definitions ($minion_job, $scheduled_product, $scheduling
       unless my $scheduled_product = $app->schema->resultset('ScheduledProducts')->find($scheduled_product_id);
     return undef unless _download_scenario_definitions($minion_job, $scheduled_product, $scheduling_params);
-    $minion_job->finish($scheduled_product->schedule_iso($scheduling_params));
+    $minion_job->finish(
+        $scheduled_product->schedule_iso($scheduling_params, $ensure_task_retry_on_termination_signal_guard));
```
```suggestion
    $minion_job->finish($scheduled_product->schedule_iso($scheduling_params, $ensure_task_retry_on_termination_signal_guard));
```

Wabri replied:
Not necessary

Wabri commented on t/api/02-iso-download.t:
```diff
@@ -67,30 +72,31 @@ sub job_gru ($job_id) { $gru_dependencies->search({job_id => $job_id})->single->
 
 # Schedule download of an existing ISO
-$rsp = schedule_iso($t, {%iso, ISO_URL => 'http://localhost/openSUSE-13.1-DVD-i586-Build0091-Media.iso'});
+$rsp
+  = schedule_iso($t, {%iso, ISO_URL => 'http://localhost/openSUSE-13.1-DVD-i586-Build0091-Media.iso'}, $signal_guard);
```
```suggestion
$rsp = schedule_iso($t, {%iso, ISO_URL => 'http://localhost/openSUSE-13.1-DVD-i586-Build0091-Media.iso'}, $signal_guard);
```

Wabri commented on t/api/02-iso.t:
```diff
@@ -87,11 +92,12 @@ subtest 'group filter and priority override' => sub {
 
-    $res = schedule_iso($t,
-        {%iso, PRECEDENCE => 'original', _GROUP => 'opensuse test', _FOO => 'bar', __FOO_URL => 'bar'});
+    $res
+      = schedule_iso($t, {%iso, PRECEDENCE => 'original', _GROUP => 'opensuse test', _FOO => 'bar', __FOO_URL => 'bar'},
+        $signal_guard);
```
```suggestion
    $res = schedule_iso($t, {%iso, PRECEDENCE => 'original', _GROUP => 'opensuse test', _FOO => 'bar', __FOO_URL => 'bar'}, $signal_guard);
```

Wabri replied:
Not necessary

Wabri commented on t/api/02-iso.t:
```diff
@@ -332,12 +340,13 @@ $t->post_ok("/api/v1/isos/$iso/cancel")->status_is(200);
 
 # handle list of tests
-$res = schedule_iso($t, {%iso, TEST => 'server,kde,textmode', _OBSOLETE => 1, _FORCE_OBSOLETE => 1}, 200);
+$res
+  = schedule_iso($t, {%iso, TEST => 'server,kde,textmode', _OBSOLETE => 1, _FORCE_OBSOLETE => 1}, $signal_guard, 200);
```
```suggestion
$res = schedule_iso($t, {%iso, TEST => 'server,kde,textmode', _OBSOLETE => 1, _FORCE_OBSOLETE => 1}, $signal_guard, 200);
```

Wabri replied:
Not necessary

Wabri commented on t/api/02-iso.t:
```diff
@@ -927,15 +937,16 @@ subtest 'schedule tests correctly when changing TEST to job template name' => su
     is($res->json->{count}, 2, 'both child and parent jobs were triggered successfully');
 
-    $res = schedule_iso($t, {%iso, _GROUP_ID => '1002', TEST => 'child_variant1', _SKIP_CHAINED_DEPS => 1});
+    $res
+      = schedule_iso($t, {%iso, _GROUP_ID => '1002', TEST => 'child_variant1', _SKIP_CHAINED_DEPS => 1}, $signal_guard);
```
```suggestion
    $res = schedule_iso($t, {%iso, _GROUP_ID => '1002', TEST => 'child_variant1', _SKIP_CHAINED_DEPS => 1}, $signal_guard);
```

Wabri replied:
Not necessary

Wabri commented:
I didn't run tidy locally, but for me have declaration and assignation in separate lines are really awful. I'm going to resolve all the conversation related to that, but still for me is not tidy at all.

b10n1k commented:
💡Mock the `OpenQA::Task::SignalGuard` for all the tests. However I found that there are other tests which seems to use the real object.
💡Update schedule_iso to reflect the new parameter
💡schedule_iso with signature now (there are other functions without the new signature definition but I didnt touch them)

Martchus commented on t/39-scheduled_products-table.t:
```diff
@@ -32,23 +33,23 @@ my %settings = (
 );
 
-# prevent job creation
+my $signal_guard = Test::MockObject->new->mock(retry => sub { undef });
+
 my $scheduled_products_mock = Test::MockModule->new('OpenQA::Schema::Result::ScheduledProducts');
 $scheduled_products_mock->redefine(_generate_jobs => sub { return undef; });
```
I'd preserve the comment and avoid the blank line.

```suggestion
my $signal_guard = Test::MockObject->new->mock(retry => sub { undef });
my $scheduled_products_mock = Test::MockModule->new('OpenQA::Schema::Result::ScheduledProducts');
$scheduled_products_mock->redefine(_generate_jobs => undef);  # prevent job creation
```

Martchus commented on t/lib/OpenQA/Test/Utils.pm:
```diff
@@ -607,7 +607,7 @@ sub mock_io_loop (%args) {
 
 sub assume_all_assets_exist { OpenQA::Schema->singleton->resultset('Assets')->search({})->update({size => 0}) }
 
-sub schedule_iso ($t, $args, $status = 200, $query_params = {}, $msg = undef) {
```
You're adding a new parameter here but it isn't used. So I suggest you leave that change out.

Martchus commented on lib/OpenQA/Schema/Result/ScheduledProducts.pm:
```diff
@@ -192,19 +192,24 @@ from B<_generate_jobs()>. Returns a list of job ids from the jobs that were succ
 
+=item $guard
+
+Expects an C<OpenQA::Task::SignalGuard> object. If database transcation fails during the scheduling,
+this ensures that the minion job will be restarted.
```
First a spelling/grammar nitpick:

```suggestion
Expects an C<OpenQA::Task::SignalGuard> object. If the database transaction fails during the scheduling,
this ensures that the Minion job will be restarted.
```

But this explanation is also completely the opposite of what passing the `$guard` will achieve here. It achieves the following:

```suggestion
Expects an C<OpenQA::Task::SignalGuard> object. If the related Minion job is aborting after the database
transaction has completed, a restart of the Minion job is prevented so jobs aren't created twice.
```

b10n1k commented:
💡Revert all the changes in the tests which use the helper schedule_iso

Martchus commented on t/39-scheduled_products-table.t:
```diff
@@ -32,23 +33,22 @@ my %settings = (
+my $signal_guard = Test::MockObject->new->mock(retry => sub { undef });
 my $scheduled_products_mock = Test::MockModule->new('OpenQA::Schema::Result::ScheduledProducts');
-$scheduled_products_mock->redefine(_generate_jobs => sub { return undef; });
+$scheduled_products_mock->redefine(_generate_jobs => sub { return undef; });    # prevent job creation
```
```suggestion
$scheduled_products_mock->redefine(_generate_jobs => undef);    # prevent job creation
```

Martchus reviewed (approved):
You'll need to resolve conflicts, though. I also have one small improvement you might want to add since you'll have to push again anyway.

Martchus commented:
I guess now only tidy complains.
